### PR TITLE
Bloom shipper: Restructure bloom store

### PR DESCRIPTION
**Pull request by chaudum**
**What this PR does / why we need it**:

This PR changes the structure of how the bloom client

* The `BloomStore` is the top component interface which implements the `Store` and the `Client` interfaces. The store holds a store entry for each schema period.
* The `bloomStoreEntry` implements the `Store` and `Client` interfaces. It holds a bloom client for a single schema period. Additionally, the store entry also exposes a fetcher for metas (and in the future for blocks), which is responsible for getting data from cache or storage (if not available in cache).
* The `BloomClient` implement the `Client` interface. The bloom client uses an object client for the schema period of the bloom client.
* The `Fetcher` can fetch and cache metas.

This structure is very similar to what we use for the chunk store.

**Note**

Before implementing `FetchBlocks()` in the `BloomStore`, I want to implement the new `FingerprintBounds` type also in the shipper and bloom gateway code to be consistent across components, as well as to make use of the new utilities exposed by them.

The store implementation probably needs some specific test cases around using the correct store entry for different schema period configs. The code however is mostly taken from the chunk store implementation.

### CODE REVIEW

File: pkg/storage/stores/shipper/bloomshipper/store.go [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/loki/pkg/storage"
+	"github.com/grafana/loki/pkg/storage/chunk/cache"
+	"github.com/grafana/loki/pkg/storage/chunk/client"
+	"github.com/grafana/loki/pkg/storage/config"
+)
+
+type Store interface {
+	SearchMetas(ctx context.Context, params MetaSearchParams) ([]Meta, error)
```
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/store.go:21]: Eventually, this should become:

```suggestion
	ResolveMetas(ctx context.Context, params MetaSearchParams) ([]MetaRef, error)
```

To match with

https://github.com/grafana/loki/blob/main/pkg/bloomcompactor/v2_meta.go#L125-L129

File: pkg/storage/stores/shipper/bloomshipper/store.go [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+		for _, object := range list {
+			metaRef, err := createMetaRef(object.Key, params.TenantID, table)
+
+			if err != nil {
+				return nil, err
+			}
+			if metaRef.MaxFingerprint < uint64(params.Keyspace.Min) || uint64(params.Keyspace.Max) < metaRef.MinFingerprint ||
+				metaRef.EndTimestamp.Before(params.Interval.Start) || metaRef.StartTimestamp.After(params.Interval.End) {
+				continue
+			}
```
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/store.go:56]: Use `Overlap()` from https://github.com/grafana/loki/blob/main/pkg/storage/bloom/v1/bounds.go#L50-L52

File: pkg/storage/stores/shipper/bloomshipper/store.go [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+			refs = append(refs, metaRef)
+		}
+	}
+	return b.fetcher.FetchMetas(ctx, refs)
```
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/store.go:60]: Should this be done in a separate function call?

```go
refs, fetchers, err := store.SearchMetas(ctx, params)
for i := range refs {
    metas, err := fetchers[i].FetchMetas(ctx, refs[i])
}
```

File: pkg/storage/stores/shipper/bloomshipper/store.go [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+}
+
+func (b *BloomStore) storeDo(ts model.Time, f func(s *bloomStoreEntry) error) error {
+	if store := b.getStore(ts); store != nil {
+		return f(store)
+	}
+	return nil
+}
+
+func (b *BloomStore) forStores(ctx context.Context, interval Interval, f func(innerCtx context.Context, interval Interval, store Store) error) error {
```
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/store.go:309]: Taken from the chunk store implementation.

File: pkg/storage/stores/shipper/bloomshipper/client.go [hunk 5]
```diff
@@ -75,7 +73,7 @@ type MetaSearchParams struct {
 type MetaClient interface {
 	// Returns all metas that are within MinFingerprint-MaxFingerprint fingerprint range
 	// and intersect time period from StartTimestamp to EndTimestamp.
-	GetMetas(ctx context.Context, params MetaSearchParams) ([]Meta, error)
+	GetMetas(ctx context.Context, metas []MetaRef) ([]Meta, error)
```
Comment (chaudum) [inline pkg/storage/stores/shipper/bloomshipper/client.go:76]: Should probably be

```suggestion
	GetMetas(ctx context.Context, metas []Meta) ([][]Meta, []Fetcher, error)
```

to adhere with the chunk client interface.

Comment (owen-d) [reply pkg/storage/stores/shipper/bloomshipper/client.go:76]: Yeah, there's some consistency there, although I think I prefer separate structures for items vs their references, as it makes it very clear what's expected of the data type (no confusion about whether it's loaded yet, etc)

File: pkg/storage/stores/shipper/bloomshipper/client.go [hunk 6]
```diff
@@ -102,84 +100,40 @@ type Client interface {
...
-		}
-		periodicObjectClients[periodicConfig.From] = objectClient
-	}
-	return &BloomClient{
-		periodicConfigs:       periodicConfigs,
-		storageConfig:         storageConfig,
-		periodicObjectClients: periodicObjectClients,
-	}, nil
-}
+var _ Client = &BloomClient{}
```
Comment (owen-d) [inline pkg/storage/stores/shipper/bloomshipper/client.go:103]: Nit: this isn't immediately clear to everyone that we're using this as a compiler check.

File: pkg/storage/stores/shipper/bloomshipper/store.go (hunk 3 above)
Comment (owen-d) [reply pkg/storage/stores/shipper/bloomshipper/store.go:60]: Yeah I'd keep these separate. Resolving Refs independently of fetching them allows us to do things without needing to download the underlying object. I think that's what you mean?

**Review by owen-d — APPROVED** [review]
Nice, thanks for this

File: pkg/storage/stores/shipper/bloomshipper/store.go (hunk 3 above)
Comment (chaudum) [reply pkg/storage/stores/shipper/bloomshipper/store.go:60]: Right, that's what I meant.

File: pkg/storage/stores/shipper/bloomshipper/client.go (hunk 6 above)
Comment (chaudum) [reply pkg/storage/stores/shipper/bloomshipper/client.go:103]: Will add a comment.

**Pull request merged**